examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,242 @@
+ */
+module ics20 {
+  import base.* from "./base"
+  import bank.* from "./bank"
+  import denomTrace.* from "./denomTrace"
+
+  /* ***************************************************************************
+   * TYPES
+   * **************************************************************************/
+
+  // Fundamental types
+  type Height = int
+
+  // IBC packet types
+  type FungibleTokenPacketAcknowledgement = {
+    success: bool,
+    errorMessage: str
+  }
+
+  type FungibleTokenData = {
+    denom: DenomTrace,
+    amount: UINT256,
+    sender: Address,
+    receiver: Address,
+    memo: str
+  }
+
+  type Packet = {
+    data: FungibleTokenData,
+    sourcePort: str,
+    sourceChannel: Channel,
+    destPort: str,
+    destChannel: Channel,
+  }
+
+  /// State of the IBC module in a chain
+  type ChainState = {
+    bank: Accounts,
+    channels: Channels,
+    outPackets: Set[Packet],
+    channelEscrowAddresses: Channel -> Address,
+  }
+
+  /// `onRecvPacket` should return an acknowledgment, but it also has to update
+  /// the chain state, so the return type is the updated value for the chain
+  /// state and the acknowledgment
+  type ReceiveTokenPacketResult = {
+    chainState: ChainState,
+    acknowledgement: FungibleTokenPacketAcknowledgement
+  }
+
+  /// The counterparty for a channel `C` in a chain is the channel identifier of
+  /// the channel `C` connects to, in the other chain.
+  pure def getCounterparty(chainState: ChainState, sourceChannel: Channel): Channel = {
+    chainState.channels.get(sourceChannel)
+  }
+
+  pure def sendFungibleTokens(chainState: ChainState, denomination: DenomTrace, amount: UINT256,
+                              sender: Address, receiver: Address, sourcePort: str, sourceChannel: Channel,
+                              timeoutHeight: Height, timeoutTimestamp: uint64): ChainState = {
+    pure val bankResult = if (movingBackAlongTrace({port: sourcePort, channel: sourceChannel}, denomination)) {
+      // burn vouchers
+      (chainState.bank).BurnCoins(sender, denomination, amount)
+    } else {
+      // escrow tokens
+      pure val escrowAccount = chainState.channelEscrowAddresses.get(sourceChannel)
+      chainState.bank.TransferCoins(sender, escrowAccount, denomination, amount)
+    }
+
+    if (bankResult.success) {
+      pure val data = { denom: denomination, amount: amount, sender: sender, receiver: receiver, memo: "" }
+
+      // handler.sendPacket
+      pure val packet = {
+        data: data,
+        sourcePort: sourcePort,
+        sourceChannel: sourceChannel,
+        destPort: "transfer",
+        destChannel: getCounterparty(chainState, sourceChannel)
+      }
+
+      chainState
+        .with("bank", bankResult.accounts)
+        .with("outPackets", chainState.outPackets.union(Set(packet)))
+    } else {
+      chainState
+    }
+  }
+
+  pure def onRecvPacket(chainState: ChainState, packet: Packet): ReceiveTokenPacketResult = {
```

Comment:
I wonder whether instead of `ReceiveTokenPacketResult` returning a pair `(Chainstate, FungibleTokenPacketAcknowledgement)` wouldn't be more readable. It becomes more clear that the intention is to return and Ack and have a side effect. What do you think?

Reply:
Makes sense! Will do